README.md
## gotestspace

gotestspace is used to quickly create a working directory for shell execution using go, as well as a tool for customizing the execution of the shell. It can help you quickly create an independent workspace for unit testing, improve the efficiency of unit test writing, and improve the happiness of developers.		

When Go projects are closely related to Linux and need to use bash shell to do some initialization or query functions, Go projects will be troublesome and repetitive to write `exec.Command` . The purpose of gotestspace is to quickly create test temp directories and execute various bash commands or shell scripts in the test temp directories to help improve the efficiency of writing unit tests.	

The main entry method of the program `testspace.Create`, when this method is called, creates a shell or a workspace that the test needs to use, in which a user-defined shell is executed and the caller can use the shell directly to initialize the repository. This method supports the following options:

* WithPathOption: Used to specify the shell working directory to be executed, you need to specify an empty directory, goshellhelper will create a directory on it, and shell execution will be done in the directory
* WithTemplateOption: This is the default shell template, which provides the shell method for test_tick by default. When this template parameter is specified, a new template will be added as an append and will not override the test_tick method. It is recommended to put the public methods in this template

```shell
test_tick () {
        if test -z "${test_tick+set}"
        then
                test_tick=1112911993
        else
                test_tick=$(($test_tick + 60))
        fi
        GIT_COMMITTER_DATE="$test_tick -0700"
        GIT_AUTHOR_DATE="$test_tick -0700"
        export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
}
```

**Note: test_tick: This method helps us reset the creator and committer time when creating a repository commit, ensuring that the commit number is the same and facilitating a quick git side test**

* WithEnvironmentsOption: Used to add custom environment variables, if not specified, then several environment variables will be provided by default:

```shell
GIT_AUTHOR_EMAIL=author@example.com
GIT_AUTHOR_NAME='A U Thor'
GIT_COMMITTER_EMAIL=committer@example.com
GIT_COMMITTER_NAME='C O Mitter'
```

* WithShellOption: user-defined shell to be executed, such as shell code for initializing the repository, etc.



When using Create directly, a temporary directory is created and the specified shell command is executed, and there are two ways to execute the shell in the already created workspace object:

* Execute: Execute the shell directly and return the standard output, standard errors, and error from go execution, for simple command execution

* ExecuteWithStdin: returns a pointer to a command object, which implements io.



During the creation process, gotestspace also adds a .git directory to the temporary directory to prevent dangerous git commands (such as git reset) from affecting the current project in the test area, so feel free to use it.

### How to use

* blackbox_test.go

As the naming of the file, this is a black-box test, also in the Go project, the namespace is not consistent with the current project, one more '_test', then here can only call the Go when the export method, so it is called as a black-box unit test for Go.

```go
package testspace_test

import (
	"context"
	"os"
	"path"
  ......
```

This file contains basic usage methods that can be used as a reference on how to use gotestspace

* example Directory

This directory shows how to integrate unit tests in Go, using techniques such as TestMain and SubTest

CHANGELOG.md
## Change Log
### [1.0.0] - 2021-12-30
The gotestspace offical release  
### ADD
* Add internal environments: CALLER and CALLER_DIR, these two environments will hold the file name and dirname for source file while call `testspace.Create()`
* Add GetMultiPath method for join multiple dir names.
* Add `Cleaner` and `Cleaner` custom method, it will be run while gotestspace run `cleanup`.
* Add gotestspace Error type, it wraps the gotestspace common error, you can find stdout and stderr on it.

### Changed
* Update the command type with interface, adds better scalability.

### Fixed
* Fix the `Cleaner`  does not work bug
* Fix internal output and outErr doesn't have value bug


helper_test.go
package testspace

import (
	"context"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestExecuteCommand(t *testing.T) {
	assert := assert.New(t)

	// Read system env
	os.Setenv("GOSHELLHELPERTEST", "Just_for_test")
	defer os.Unsetenv("GOSHELLHELPERTEST")

	cancelCtx1, cancel1 := context.WithCancel(context.Background())
	defer cancel1()
	output, _, err := SimpleExecuteCommand(cancelCtx1, "", nil, "env")
	assert.NoError(err)
	assert.Contains(output, "GOSHELLHELPERTEST=Just_for_test")

	output, _, err = SimpleExecuteCommand(cancelCtx1, "",
		[]string{"GOSHELLHELPERTEST=Just_for_test123"},
		"env")
	assert.NoError(err)
	assert.Contains(output, "GOSHELLHELPERTEST=Just_for_test123")

	_, _, err = SimpleExecuteCommand(cancelCtx1, "", nil, "evn1111")
	if err == nil {
		assert.Error(err)
	}

	_, stderr, err := SimpleExecuteCommand(cancelCtx1, "", nil, "git", "aaa")
	assert.Error(err)
	assert.Contains(stderr, "git --help")
}


blackbox_test.go
package testspace_test

import (
	"bytes"
	"context"
	"os"
	"path"
	"strings"
	"testing"

	testspace "github.com/Jiu2015/gotestspace"
	"github.com/stretchr/testify/assert"
)

// The sample test for running "printf hello"
func TestSampleShell(t *testing.T) {
	assert := assert.New(t)
	workspace, err := testspace.Create(testspace.WithShellOption("printf \"hello\""))
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}
	defer workspace.Cleanup()

	assert.Equal("hello", workspace.GetOutputStr())
}

// Add environment example
func TestSampleShellWithEnvironments(t *testing.T) {
	assert := assert.New(t)
	workspace, err := testspace.Create(
		// Add two environments Testing1 and Testing2
		testspace.WithEnvironmentsOption("Testing1=aa", "Testing2=bb"),
		testspace.WithShellOption("printf \"%s, %s\" $Testing1 $Testing2"),
	)
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}
	defer workspace.Cleanup()

	assert.Equal("aa, bb", workspace.GetOutputStr())
}

// Add template example
func TestAddTemplateAndCall(t *testing.T) {
	assert := assert.New(t)
	workspace, err := testspace.Create(
		testspace.WithTemplateOption(`
test(){
	printf "this is a test from test method"
}
`),
		testspace.WithShellOption("test"))
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}
	defer workspace.Cleanup()

	assert.Equal("this is a test from test method", workspace.GetOutputStr())
}

// Add custom path example
func TestSetCustomPathForTesting(t *testing.T) {
	assert := assert.New(t)
	currentPath, _ := os.Getwd()
	testFolderName := "testing_folder"
	testPath := path.Join(currentPath, testFolderName)
	workspace, err := testspace.Create(
		testspace.WithPathOption(testPath),
	)
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}
	defer workspace.Cleanup()

	workPath := workspace.GetPath(testFolderName)

	_, err = os.Stat(workPath)
	if err != nil {
		assert.Error(err, "the custom path not exist")
	}
}

// Create a bare repository example
func TestCreateBareRepository(t *testing.T) {
	assert := assert.New(t)

	// The "test_tick" is the default method in template
	workspace, err := testspace.Create(
		testspace.WithShellOption(`
git init --bare test.git &&
git clone test.git test && 
(
	cd test && 
	printf "this is a test\n">init.js &&
	git add init.js &&
	test_tick &&
	git commit -m "this is the first commit" &&
	git push 
) &&
rm -rf test
`))
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%s", workspace.GetOutputStr())
	}
	defer workspace.Cleanup()

	cancelCtx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Let's add the second commit, running custom shell again
	_, _, err = workspace.Execute(cancelCtx, `
git clone test.git test && 
(
	cd test && 
	printf "add a new file\n"> main.go &&
	git add main.go && 
	test_tick &&
	git commit -m "this is the second commit" && 
	git push 
)&&
rm -rf test
`)
	if !assert.NoError(err, "create testspace got error") {
		assert.FailNowf("create testspace got error", "%s", workspace.GetOutputStr())
	}

	// Now, let's check the bare repository
	_, _, err = workspace.Execute(cancelCtx, "cd test.git && git log --oneline")
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}

	assert.Equal("95dbed8 this is the second commit\n5a1f64b this is the first commit",
		strings.TrimSpace(workspace.GetOutputStr()))
}

// Test use stdin to delete bare repository branch
func TestDeleteBareRepositoryBranch(t *testing.T) {
	assert := assert.New(t)
	// The "test_tick" is the default method in template
	workspace, err := testspace.Create(
		testspace.WithShellOption(`
git init --bare test.git &&
git clone test.git test && 
(
	cd test && 
	printf "this is a test">init.js &&
	git add init.js &&
	test_tick &&
	git commit -m "this is the first commit" &&
	git push &&
	printf "the js file content">resource.js &&
	git add resource.js &&
	test_tick &&
	git commit -m "the second commit" && 
	git branch branch1 &&
	printf "the go file">main.go &&
	git add main.go &&
	test_tick &&
	git commit -m "the third commit" &&
	git branch branch2 &&
	git push --all
) &&
rm -rf test
`))
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}
	defer workspace.Cleanup()

	cancelCtx, cancel := context.WithCancel(context.Background())
	defer cancel()

	command, err := workspace.ExecuteWithStdin(cancelCtx, "cd test.git && git update-ref --stdin")
	if !assert.NoError(err) {
		assert.FailNowf("execute with stdin got error", "%v", err)
	}

	// Start delete branch1
	command.Write([]byte("delete refs/heads/branch1\n"))

	// Start delete branch2
	command.Write([]byte("delete refs/heads/branch2\n"))

	if err = command.Wait(); !assert.NoError(err) {
		assert.FailNowf("execute with stdin got error", "%v", err)
	}

	// Let's check all the branches
	stdout, _, err := workspace.Execute(cancelCtx, "cd test.git && git branch --format=\"%(refname)\"")
	if !assert.NoError(err) {
		assert.FailNowf("execute git command got error", "%s", err)
	}

	assert.Equal("refs/heads/master\n", stdout)
}

// Test use stdin to delete bare repository branch
func TestDeleteBareRepositoryBranchByStdinAndStdoutCommand(t *testing.T) {
	var (
		stdin  bytes.Buffer
		stdout bytes.Buffer
	)

	assert := assert.New(t)

	// The "test_tick" is the default method in template
	workspace, err := testspace.Create(
		testspace.WithShellOption(`
git init --bare test.git &&
git clone test.git test && 
(
	cd test && 
	printf "this is a test">init.js &&
	git add init.js &&
	test_tick &&
	git commit -m "this is the first commit" &&
	git push &&
	printf "the js file content">resource.js &&
	git add resource.js &&
	test_tick &&
	git commit -m "the second commit" && 
	git branch branch1 &&
	printf "the go file">main.go &&
	git add main.go &&
	test_tick &&
	git commit -m "the third commit" &&
	git branch branch2 &&
	git push --all
) &&
rm -rf test
`))
	if !assert.NoError(err) {
		assert.FailNowf("create testspace got error", "%v", err)
	}
	defer workspace.Cleanup()

	cancelCtx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Start delete branch1.
	// It needs before command execute.
	stdin.Write([]byte("delete refs/heads/branch1\n"))

	// Start delete branch2
	// It needs before command execute.
	stdin.Write([]byte("delete refs/heads/branch2\n"))

	command, err := workspace.ExecuteWithStdinStdout(cancelCtx, nil, &stdin, &stdout,
		"cd test.git && git update-ref --stdin")
	if !assert.NoError(err) {
		assert.FailNowf("execute with stdin got error", "%v", err)
	}

	if err = command.Wait(); !assert.NoError(err) {
		assert.FailNowf("execute with stdin got error", "%v", err)
	}

	// Let's check all the branches
	outStr, _, err := workspace.Execute(cancelCtx, "cd test.git && git branch --format=\"%(refname)\"")
	if !assert.NoError(err) {
		assert.FailNowf("execute git command got error", "%s", err)
	}

	assert.Equal("refs/heads/master\n", outStr)
	assert.Equal("", stdout.String())
}


testspace_test.go
package testspace

import (
	"errors"
	"fmt"
	"os"
	"path"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNewShellSpace(t *testing.T) {
	currentPath, _ := os.Getwd()

	type args struct {
		options []CreateOption
	}
	tests := []struct {
		name    string
		args    args
		want    Space
		wantErr bool
	}{
		{
			name: "only_path_parameter",
			args: args{
				options: []CreateOption{
					WithPathOption("tmp"),
				},
			},
			want: &workSpace{
				path: path.Join(currentPath, "tmp"),
				env: []string{
					fmt.Sprintf("HOME=%s/tmp", currentPath),
					"GIT_AUTHOR_EMAIL=author@example.com",
					"GIT_AUTHOR_NAME='A U Thor'",
					"GIT_COMMITTER_EMAIL=committer@example.com",
					"GIT_COMMITTER_NAME='C O Mitter'",
				},
				template: `
test_tick () {
        if test -z "${test_tick+set}"
        then
                test_tick=1112911993
        else
                test_tick=$(($test_tick + 60))
        fi
        GIT_COMMITTER_DATE="$test_tick -0700"
        GIT_AUTHOR_DATE="$test_tick -0700"
        export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
}
`,
				customShell: "",
				output:      "",
			},
			wantErr: false,
		},
		{
			name: "path_and_env_parameter",
			args: args{
				options: []CreateOption{
					WithPathOption("tmp"),
					WithEnvironmentsOption("goshelltest1=111", "goshelltest2=222"),
				},
			},
			want: &workSpace{
				path: path.Join(currentPath, "tmp"),
				env: []string{
					fmt.Sprintf("HOME=%s/tmp", currentPath),
					"GIT_AUTHOR_EMAIL=author@example.com",
					"GIT_AUTHOR_NAME='A U Thor'",
					"GIT_COMMITTER_EMAIL=committer@example.com",
					"GIT_COMMITTER_NAME='C O Mitter'",
					"goshelltest1=111",
					"goshelltest2=222",
				},
				template: `
test_tick () {
        if test -z "${test_tick+set}"
        then
                test_tick=1112911993
        else
                test_tick=$(($test_tick + 60))
        fi
        GIT_COMMITTER_DATE="$test_tick -0700"
        GIT_AUTHOR_DATE="$test_tick -0700"
        export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
}
`,
				customShell: "",
				output:      "",
			},
			wantErr: false,
		},
		{
			name: "path_and_env_and_template_parameter",
			args: args{
				options: []CreateOption{
					WithPathOption("tmp"),
					WithEnvironmentsOption("goshelltest1=111", "goshelltest2=222"),
					WithTemplateOption(`
test(){
	echo hello
}
`),
				},
			},
			want: &workSpace{
				path: path.Join(currentPath, "tmp"),
				env: []string{
					fmt.Sprintf("HOME=%s/tmp", currentPath),
					"GIT_AUTHOR_EMAIL=author@example.com",
					"GIT_AUTHOR_NAME='A U Thor'",
					"GIT_COMMITTER_EMAIL=committer@example.com",
					"GIT_COMMITTER_NAME='C O Mitter'",
					"goshelltest1=111",
					"goshelltest2=222",
				},
				template: `
test_tick () {
        if test -z "${test_tick+set}"
        then
                test_tick=1112911993
        else
                test_tick=$(($test_tick + 60))
        fi
        GIT_COMMITTER_DATE="$test_tick -0700"
        GIT_AUTHOR_DATE="$test_tick -0700"
        export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
}


test(){
	echo hello
}
`,
				customShell: "",
				output:      "",
			},
			wantErr: false,
		},
		{
			name: "path_and_env_and_template_and_customshell_parameter",
			args: args{
				options: []CreateOption{
					WithPathOption("tmp"),
					WithEnvironmentsOption("goshelltest1=111", "goshelltest2=222"),
					WithTemplateOption(`
test(){
	echo hello
}
`),
					WithShellOption("test"),
				},
			},
			want: &workSpace{
				path: path.Join(currentPath, "tmp"),
				env: []string{
					fmt.Sprintf("HOME=%s/tmp", currentPath),
					"GIT_AUTHOR_EMAIL=author@example.com",
					"GIT_AUTHOR_NAME='A U Thor'",
					"GIT_COMMITTER_EMAIL=committer@example.com",
					"GIT_COMMITTER_NAME='C O Mitter'",
					"goshelltest1=111",
					"goshelltest2=222",
				},
				template: `
test_tick () {
        if test -z "${test_tick+set}"
        then
                test_tick=1112911993
        else
                test_tick=$(($test_tick + 60))
        fi
        GIT_COMMITTER_DATE="$test_tick -0700"
        GIT_AUTHOR_DATE="$test_tick -0700"
        export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
}


test(){
	echo hello
}
`,
				customShell: "test",
				output:      "hello\n",
			},
			wantErr: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			actual, err := Create(tt.args.options...)
			if (err != nil) != tt.wantErr {
				t.Errorf("Create() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			defer actual.Cleanup()

			assert.Equal(t, tt.want.GetPath(""), actual.GetPath(""))
			assert.Equal(t, tt.want.GetTemplateStr(), actual.GetTemplateStr())
			envs1 := tt.want.GetEnvStr()
			envs2 := actual.GetEnvStr()
			if assert.True(t,
				len(envs1) <= len(envs2),
				"actual envs less than expect envs") {
				for i := range envs1 {
					assert.Equal(t, envs1[i], envs2[i])
				}

			}
			_, err = os.Stat(actual.GetPath(".git"))
			if err != nil {
				t.Error("git init failed")
			}
		})
	}
}

func Test_workSpace_RegistrationCustomCleaner(t *testing.T) {
	type fields struct {
		path        string
		env         []string
		template    string
		customShell string
		output      string
		outErr      string
		cleaners    []CustomCleaner
	}
	type args struct {
		cleaners []CustomCleaner
	}
	tests := []struct {
		name        string
		fields      fields
		args        args
		expectCount int
	}{
		{
			name: "add_nil",
			fields: fields{
				cleaners: nil,
			},
			args: args{
				cleaners: nil,
			},
			expectCount: 0,
		}, {
			name:   "add_one",
			fields: fields{},
			args: args{
				cleaners: []CustomCleaner{
					func() error {
						return nil
					},
				},
			},
			expectCount: 1,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			w := &workSpace{
				path:        tt.fields.path,
				env:         tt.fields.env,
				template:    tt.fields.template,
				customShell: tt.fields.customShell,
				output:      tt.fields.output,
				outErr:      tt.fields.outErr,
				cleaners:    tt.fields.cleaners,
			}
			w.RegistrationCustomCleaner(tt.args.cleaners...)

			if len(w.cleaners) != tt.expectCount {
				t.Errorf("the cleanner count invalid, expected: %d, acture: %d", tt.expectCount, len(w.cleaners))
			}
		})
	}
}

func Test_workSpace_Cleanup(t *testing.T) {
	tmpPath := func() string {
		s, err := os.MkdirTemp("", "test*")
		if err != nil {
			panic(err)
		}

		return s
	}()

	// Just used for cleaner
	var tmpNum int

	type fields struct {
		path        string
		env         []string
		template    string
		customShell string
		output      string
		outErr      string
		cleaners    []CustomCleaner
	}
	tests := []struct {
		name        string
		fields      fields
		wantErr     bool
		expectCheck func() bool
	}{
		{
			name: "no_customer_cleaner",
			fields: fields{
				path: tmpPath,
			},
			wantErr: false,
		},
		{
			name:    "no_path",
			fields:  fields{},
			wantErr: true,
		},
		{
			name: "with_cleaner",
			fields: fields{
				path: tmpPath,
				cleaners: []CustomCleaner{
					func() error {
						tmpNum = 100
						return nil
					},
				},
			},
			wantErr: false,
			expectCheck: func() bool {
				return tmpNum == 100
			},
		},
		{
			name: "with_multiple_cleaner",
			fields: fields{
				path: tmpPath,
				cleaners: []CustomCleaner{
					func() error {
						tmpNum = 100
						return nil
					},
					func() error {
						tmpNum += 100
						return nil
					},
				},
			},
			wantErr: false,
			expectCheck: func() bool {
				return tmpNum == 200
			},
		},
		{
			name: "one_cleaner_error",
			fields: fields{
				path: tmpPath,
				cleaners: []CustomCleaner{
					func() error {
						tmpNum = 100
						return nil
					},
					func() error {
						return errors.New("something went wrong")
					},
				},
			},
			wantErr:     true,
			expectCheck: nil,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			w := &workSpace{
				path:        tt.fields.path,
				env:         tt.fields.env,
				template:    tt.fields.template,
				customShell: tt.fields.customShell,
				output:      tt.fields.output,
				outErr:      tt.fields.outErr,
				cleaners:    tt.fields.cleaners,
			}
			if err := w.Cleanup(); err != nil != tt.wantErr {
				t.Errorf("got error: %v", err)
			}

			if tt.expectCheck != nil {
				if !tt.expectCheck() {
					t.Error("the cleaner invalid")
				}
			}
		})
	}
}


errors.go
package testspace

import "fmt"

const (
	// GenerateError the generate error, the exit code is 1
	GenerateError = 1

	// Misuse the misuse command, the exit code is 2
	Misuse = 2

	// NotExecute the not execute command error, the exit code is 126
	NotExecute = 126

	// NotFound the command not found error, the exit code is 127
	NotFound = 127

	// InvalidArgument the command has invalid argument, the exit code is 128
	InvalidArgument = 128
)

// Error gotestspace custom error
type Error struct {
	// The exit code with command
	exitCode int

	// The exit with signal, if exit but not through system singal, it will be 0
	signal int

	// The original error
	err error

	// The standard output
	stdout string

	// The standard error output
	stderr string
}

// Error print the generate error
func (e Error) Error() string {
	if e.err != nil {
		return e.err.Error()
	}

	return ""
}

// Unwrap get the origin error
func (e Error) Unwrap() error {
	return e.err
}

// GetMessage get the message from custom error
// it will print the error from exitcode and signal
func (e *Error) GetMessage() string {
	if e.err == nil {
		return ""
	}

	if e.signal > 0 {
		return fmt.Sprintf("command interrupt by singal: %d", e.signal)
	}

	switch e.exitCode {
	case GenerateError:
		return "catch shell generate error"
	case Misuse:
		return "misuse of shell build-in"
	case NotExecute:
		return "command cannot execute"
	case NotFound:
		return "command cannot found"
	case InvalidArgument:
		return "invalid argument"
	default:
		return "unknown error"
	}
}

// GetStdout get the standard output
func (e *Error) GetStdout() string {
	return e.stdout
}

// GetStderr get the standard error output
func (e *Error) GetStderr() string {
	return e.stderr
}

// GetExitCode get the exit code
func (e *Error) GetExitCode() int {
	return e.exitCode
}

// NewGoTestSpaceError create new gotestspace error
func NewGoTestSpaceError(exitCode int, stdout, stderr string, err error) Error {
	signal := splitCodeWithSignal(exitCode)

	return Error{
		exitCode: exitCode,
		signal:   signal,
		stdout:   stdout,
		stderr:   stderr,
		err:      err,
	}
}

// splitCodeWithSignal used to check exit code and get the signal number
func splitCodeWithSignal(exitCode int) (signal int) {
	if exitCode > 128 && exitCode < 255 {
		return exitCode - 128
	}

	return 0
}


helper.go
package testspace

import (
	"bytes"
	"context"
	"errors"
	"io"
	"io/ioutil"
	"os/exec"
)

// SimpleExecuteCommand to execute cmd sample method.
// The error may be testspace.Error type, so please use type assertions or use errors.As
func SimpleExecuteCommand(ctx context.Context, path string, env []string, commandName string,
	args ...string) (output string, outErr string, err error) {
	cmd := exec.Command(commandName, args...)

	if len(path) > 0 {
		cmd.Dir = path
	}

	stderr := &bytes.Buffer{}

	spaceCommand, err := new(ctx, cmd, nil, nil, stderr, env...)
	if err != nil {
		return "", stderr.String(), err
	}

	cmdStdout, err := ioutil.ReadAll(spaceCommand)
	if err != nil {
		return "", stderr.String(), err
	}

	output = string(cmdStdout)
	outErr = spaceCommand.GetStderr()

	if err = cmd.Wait(); err != nil {
		var exitError *exec.ExitError
		if errors.As(err, &exitError) {
			exitError.ExitCode()
			// Wrap error to testspace.Error
			err = NewGoTestSpaceError(exitError.ExitCode(), output, stderr.String(), err)
		}

		return output, stderr.String(), err
	}

	return output, outErr, err
}

// NewTestSpaceCommand will return space-command for advantage use,
// You must get stdin, stdout and stderr before spaceCommand.Wait(), and do not miss spaceCommand.Wait()
func NewTestSpaceCommand(ctx context.Context, path string, env []string, stdin io.Reader, stdout, stderr io.Writer,
	commandName string, args ...string) (Commander, error) {
	cmd := exec.Command(commandName, args...)

	if len(path) > 0 {
		cmd.Dir = path
	}

	spaceCommand, err := new(ctx, cmd, stdin, stdout, stderr, env...)
	if err != nil {
		return nil, err
	}

	return spaceCommand, nil
}


command.go
package testspace

import (
	"context"
	"errors"
	"io"
	"io/ioutil"
	"os/exec"
	"sync"
	"syscall"
)

// Commander use execute command
type Commander interface {
	Read(p []byte) (int, error)
	Write(p []byte) (int, error)
	Wait() error
	GetStderr() string
}
type command struct {
	reader   io.Reader
	writer   io.WriteCloser
	stderr   *stdErr
	cmd      *exec.Cmd
	context  context.Context
	waitOnce sync.Once
	waitErr  error
}

func (c *command) Read(p []byte) (int, error) {
	if c.reader == nil {
		panic("command invalid with reader")
	}

	return c.reader.Read(p)
}

func (c *command) Write(p []byte) (int, error) {
	if c.writer == nil {
		panic("command invalid with writer")
	}

	return c.writer.Write(p)
}

func (c *command) wait() {
	if c.writer != nil {
		c.writer.Close()
	}

	if c.reader != nil {
		io.Copy(ioutil.Discard, c.reader)
	}

	c.waitErr = c.cmd.Wait()
}

// Wait waits the command exit and wait the stdin, stdout
// The error may be testspace.Error type, so please use type assertions or use errors.As
func (c *command) Wait() error {
	var (
		err       error
		exitError *exec.ExitError
	)

	c.waitOnce.Do(c.wait)

	err = c.waitErr
	if errors.As(err, &exitError) {
		exitError.ExitCode()
		// Wrap error to testspace.Error
		err = NewGoTestSpaceError(exitError.ExitCode(), "", string(c.stderr.GetStderr()), err)
	}

	return err
}

// GetStderr get command stderr
func (c *command) GetStderr() string {
	if c.stderr == nil {
		return ""
	}

	return string(c.stderr.GetStderr())
}

// The command use to check to set stdin type
var setStdinType io.Reader = setStdin{}

type setStdin struct{}

func (setStdin) Read([]byte) (int, error) {
	return 0, errors.New("just used for tag if or not use command stdin")
}

type stdErr struct {
	errContents []byte
}

func (s *stdErr) Write(p []byte) (int, error) {
	if len(p) > 0 {
		s.errContents = append(s.errContents, p...)
		return len(p), nil
	}

	return 0, nil
}

func (s *stdErr) GetStderr() []byte {
	return s.errContents
}

func new(ctx context.Context, cmd *exec.Cmd, stdin io.Reader, stdout, stderr io.Writer, env ...string) (Commander, error) {
	if ctx.Done() == nil {
		panic("the context must have Done() method, use WithCancel, WithTimeout etc.")
	}

	resultCommand := &command{
		cmd:     cmd,
		context: ctx,
		stderr:  &stdErr{},
	}

	if len(env) > 0 {
		cmd.Env = append(cmd.Env, env...)
	}

	if stdin == setStdinType {
		pip, err := cmd.StdinPipe()
		if err != nil {
			return nil, err
		}

		resultCommand.writer = pip
	}

	if resultCommand.writer == nil && stdin != nil {
		cmd.Stdin = stdin
	}

	if stdout == nil {
		pip, err := cmd.StdoutPipe()
		if err != nil {
			return nil, err
		}
		resultCommand.reader = pip
	} else {
		cmd.Stdout = stdout
	}

	if stderr != nil {
		cmd.Stderr = stderr
	} else {
		cmd.Stderr = resultCommand.stderr
	}

	// Set the child process have a new process group
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true,
	}

	if err := cmd.Start(); err != nil {
		return nil, err
	}

	go func() {
		<-ctx.Done()

		if process := cmd.Process; process != nil && process.Pid > 0 {
			// Kill the process and it's child process
			syscall.Kill(-process.Pid, syscall.SIGKILL)
		}
	}()

	return resultCommand, nil
}


options.go
package testspace

import (
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"path/filepath"
	"strings"
	"time"
)

// createOption the create workspace option
type createOption struct {
	workspacePath string
	template      string
	environments  []string
	customShell   string
	cleaners      []CustomCleaner
}

// CreateOption the option for create shell workspace
type CreateOption struct {
	setOpt func(opt *createOption) error
}

// WithPathOption set the path
func WithPathOption(workPath string) CreateOption {
	var (
		fullPath string
		err      error
	)
	workPath = filepath.Clean(workPath)
	basePath := filepath.Base(workPath)
	dirPath := filepath.Dir(workPath)
	if strings.Contains(basePath, "*") {
		if filepath.IsAbs(dirPath) {
			fullPath, err = ioutil.TempDir(dirPath, basePath)
		} else {
			fullPath, err = ioutil.TempDir("", workPath)
		}
	} else if !filepath.IsAbs(workPath) {
		fullPath, err = os.Getwd()
		if err == nil {
			fullPath = path.Join(fullPath, workPath)
		}
	} else {
		fullPath = workPath
	}

	return CreateOption{setOpt: func(opt *createOption) error {
		opt.workspacePath = fullPath
		if err != nil {
			return fmt.Errorf("the WithPathOption invalid: %v", err)
		}

		return nil
	}}
}

// WithTemplateOption set custom template
func WithTemplateOption(customTemplate string) CreateOption {
	return CreateOption{setOpt: func(opt *createOption) error {
		// The template will append the default template which init test_tick function
		opt.template = opt.template + "\n" + customTemplate
		return nil
	}}
}

// WithEnvironmentsOption set environments
func WithEnvironmentsOption(environments ...string) CreateOption {
	return CreateOption{setOpt: func(opt *createOption) error {
		opt.environments = append(opt.environments, environments...)
		return nil
	}}
}

// WithShellOption set custom shell from user
func WithShellOption(customShell string) CreateOption {
	return CreateOption{setOpt: func(opt *createOption) error {
		opt.customShell = customShell
		return nil
	}}
}

// WithCleanersOption set the custom cleaners
func WithCleanersOption(cleaners ...CustomCleaner) CreateOption {
	return CreateOption{
		setOpt: func(opt *createOption) error {
			opt.cleaners = append(opt.cleaners, cleaners...)
			return nil
		}}
}

func mergeOptions(options []CreateOption) *createOption {
	var err error
	currentPath, _ := os.Getwd()
	tempWorkspacePath := path.Join(currentPath, time.Now().Format("20060102150405"))
	o := &createOption{
		// Default path
		workspacePath: tempWorkspacePath,
		template: `
test_tick () {
        if test -z "${test_tick+set}"
        then
                test_tick=1112911993
        else
                test_tick=$(($test_tick + 60))
        fi
        GIT_COMMITTER_DATE="$test_tick -0700"
        GIT_AUTHOR_DATE="$test_tick -0700"
        export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
}
`,
		environments: []string{
			"GIT_AUTHOR_EMAIL=author@example.com",
			"GIT_AUTHOR_NAME='A U Thor'",
			"GIT_COMMITTER_EMAIL=committer@example.com",
			"GIT_COMMITTER_NAME='C O Mitter'",
		},
		customShell: "",
	}

	for _, opt := range options {
		err = opt.setOpt(o)
		if err != nil {
			panic(err)
		}
	}

	o.environments = append([]string{fmt.Sprintf("HOME=%s", o.workspacePath)},
		o.environments...)

	return o
}
